flink-runtime/src/main/java/org/apache/flink/runtime/minicluster/MiniClusterConfiguration.java

```diff
@@ -50,6 +51,8 @@ public class MiniClusterConfiguration {
 
     private final MiniCluster.HaServices haServices;
 
+    @Nullable private final PluginManager pluginManager;
```

Comment:
I'd have the user pass a `Path` to a plugin directory instead. This exposes less internals, ensures we can do proper lifecycle management of the plugin manager and makes it more obvious what you actually have to do to set up plugins.
Additionally this would allow us to also expose this in the MiniClusterExtension.

Reply:
I see your concern. It is indeed not intuitive how to initialize a plugin manager, and for test, just passing a directory would be easier.
My motivation for this change comes from a non-testing scenario, where I first initialize the PluginManager, then I initialize the FileSystems, and finally the MiniCluster so that it can set up the metrics reporters. In that use-case, it makes sense to have the PluginManager lifecycle managed outside of the MiniCluster, to avoid instantiating it twice.

How about adding a comment to the `MiniClusterConfiguration.Builder.setPluginManager()` method about how to initialize the PluginManager for now? This would address your concern about how to initialize it.
Once somebody has a need for using the MiniCluster with plugins in a unit test with MiniClusterExtension, we can evolve this feature further.

Reply:
I mean, you probably shouldn't be using the `MiniCluster` directly in the first place.
As far as I'm concerned the `MiniCluster` is internal; the `MiniClusterExtension` is the user-facing API.

That being said, if one considers the `MiniCluster` as internal then accepting a `PluginManager` is fine; you'll just have to pay the price for using such APIs.
(At some point) the `MiniClusterExtension` could accept a `Path` and then construct (and manage) the `PluginManager`.
I do recall users requesting this for testing actually (It's just a bit eh that we'd support reporter plugins, but not filesystems (damn singletons :anger: )).

Reply:
MiniCluster is indeed not documented, hence I agree that it is rather internal. But I know from at least one other company that has been using it directly in their production code. So maybe we should consider documenting it properly and making it a public API eventually.

But I agree with your conclusion that passing an internal API (`PluginManager`) to another internal API (`MiniCluster`) is fine.

I've pushed a commit addressing your feedback.

Reply:
> So maybe we should consider documenting it properly and making it a public API eventually.

_Eventually_, maybe, but we'd have to make a fair number of changes; it's currently exposing to way many internals to be a true user-facing thing.

We'd need a wrapper around it, similar to what the `MiniClusterExtension` does. Also, at that point passing a `PluginManager` would again be a no-go because that's also an internal API (and I don't think we should change that).